plugins/salesforce/core/src/main/java/org/pentaho/di/trans/steps/salesforceinput/SalesforceInput.java

```diff
@@ -438,4 +460,189 @@ public void dispose( StepMetaInterface smi, StepDataInterface sdi ) {
+  private JSONObject getFieldsAction() {
+    JSONObject response = new JSONObject();
+    response.put( StepInterface.ACTION_STATUS, StepInterface.FAILURE_RESPONSE );
+    try {
+      FieldsResponse fieldsResponse = getFields();
+      ObjectMapper objectMapper = new ObjectMapper();
+      String jsonString = objectMapper.writeValueAsString( fieldsResponse );
+      response.put( "fieldsResponse", objectMapper.readTree( jsonString ) );
+      response.put( StepInterface.ACTION_STATUS, StepInterface.SUCCESS_RESPONSE );
+      response.put( StepInterface.STATUS, StepInterface.SUCCESS_STATUS );
+    } catch ( Exception e ) {
+      response.put( "errorMsg", BaseMessages.getString( PKG, "SalesforceInputMeta.ErrorRetrieveData.DialogMessage" ) );
+      StringBuilder details = new StringBuilder();
+      StringWriter sw = new StringWriter();
+      PrintWriter pw = new PrintWriter( sw );
+      e.printStackTrace( pw );
+      details.append( sw.getBuffer() );
+      response.put( "details", details.toString() );
+      log.logError( e.getMessage() );
+      response.put( StepInterface.STATUS, StepInterface.FAILURE_STATUS );
+    }
+    return response;
+  }
+
+  public FieldsResponse getFields() throws Exception {
+    SalesforceConnection connection = null;
+    try {
+      SalesforceInputMeta meta = (SalesforceInputMeta) getStepMetaInterface();
+      String realURL = getTransMeta().environmentSubstitute( meta.getTargetURL() );
+      String realUsername = getTransMeta().environmentSubstitute( meta.getUsername() );
+      String realPassword = Utils.resolvePassword( getTransMeta(), meta.getPassword() );
+      int realTimeOut = Const.toInt( getTransMeta().environmentSubstitute( meta.getTimeout() ), 0 );
+
+      connection = new SalesforceConnection( log, realURL, realUsername, realPassword );
+      connection.setTimeOut( realTimeOut );
+      FieldsResponse fieldsResponse = new FieldsResponse( new ArrayList<>(), new HashSet<>() );
+      if ( meta.isSpecifyQuery() ) {
+        // Free hand SOQL
+        String realQuery = getTransMeta().environmentSubstitute( meta.getQuery() );
+        connection.setSQL( realQuery );
+        connection.connect();
+        // We are connected, so let's query
+        XmlObject[] fields = connection.getElements();
+        int nrFields = fields.length;
+        for ( int i = 0; i < nrFields; i++ ) {
+          fieldsResponse.fieldDTOList.addAll( addFields( "", fieldsResponse.fieldNames, fields[ i ] ) );
+        }
+      } else {
+        connection.connect();
+
+        Field[] fields = connection.getObjectFields( meta.getModule() );
+        for ( int i = 0; i < fields.length; i++ ) {
+          Field field = fields[ i ];
+          fieldsResponse.fieldNames.add( field.getName() );
+          fieldsResponse.fieldDTOList.add( addField( field ) );
+        }
+      }
+      return fieldsResponse;
+
+    } finally {
+      if ( connection != null ) {
+        try {
+          connection.close();
```

Comment:
Shouldn't you have the `connection.close()` in finally clause for` SalesforceInput#getFields`? If connection variable is a AutoCloseable you can wrap it in a ry-with-resource statement and java will handle all the nuance of properly closing IO https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html